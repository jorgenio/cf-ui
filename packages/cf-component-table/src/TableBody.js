import React from 'react';
import PropTypes from 'prop-types';
import { filterStyle, mapChildren } from 'cf-style-container';

class TableBody extends React.Component {
  render() {
    const {
      className,
      styles,
      striped,
      hover,
      bordered,
      condensed,
      bare,
      tbodyIndex,
      children,
      ...props
    } = this.props;

    return (
      <tbody
        className={className || (styles && styles.tableBody)}
        {...filterStyle(props)}
      >
        {mapChildren(children, (child, index, children) => {
          const isLast = index === children.length - 1;
          if (React.isValidElement(child)) {
            return React.cloneElement(child, {
              condensed,
              striped,
              bordered,
              hover,
              bare,
              tbodyIndex,
              rowIndex: isLast ? -index : index
            });
          }
          return child;
        })}
      </tbody>
    );
  }
}

TableBody.propTypes = {
  className: PropTypes.string,
  styles: PropTypes.object,
  children: PropTypes.node
};

export default TableBody;
