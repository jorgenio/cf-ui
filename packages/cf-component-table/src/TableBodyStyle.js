import { filterNone } from 'cf-style-container';

export default ({ theme, tbodyIndex }) =>
  filterNone({
    borderTop: tbodyIndex > 0
      ? '2px solid ${theme.tableBorderColor}'
      : undefined,

    tablet: {
      display: 'block',
      width: '100%',
      boxSizing: 'border-box'
    },

    [`@media screen and (max-width: ${theme.breakpoints.tablet})`]: {
      display: 'block'
    }
  });
