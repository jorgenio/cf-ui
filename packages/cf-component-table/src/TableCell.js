import React from 'react';
import PropTypes from 'prop-types';
import { filterStyle } from 'cf-style-container';

class TableCell extends React.Component {
  render() {
    const {
      className,
      styles,
      children,
      // Start ignoring, because Fela...
      condensed,
      striped,
      bordered,
      hover,
      bare,
      rowType,
      theadIndex,
      tbodyIndex,
      tfootIndex,
      rowIndex,
      cellIndex,
      align,
      // Stop ignoring...
      ...props
    } = this.props;

    return (
      <td
        className={className || (styles && styles.tableCell)}
        {...filterStyle(props)}
      >
        {this.props.children}
      </td>
    );
  }
}

TableCell.propTypes = {
  className: PropTypes.string,
  styles: PropTypes.object,
  align: PropTypes.oneOf(['left', 'center', 'right']),
  children: PropTypes.node
};

TableCell.defaultProps = {
  align: 'left'
};

export default TableCell;
