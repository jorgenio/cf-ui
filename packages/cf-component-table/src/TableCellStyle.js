import { capitalize, filterNone } from 'cf-style-container';

export default ({
  theme,
  condensed,
  striped,
  bordered,
  hover,
  bare,
  rowType,
  theadIndex,
  tbodyIndex,
  tfootIndex,
  rowIndex,
  cellIndex,
  align
}) =>
  filterNone({
    // Baseline cell styles
    borderTop: ((theadIndex === 0 || tbodyIndex === 0) && rowIndex === 0) ||
      bare
      ? 0
      : `1px solid ${theme.tableBorderColor}`,
    lineHeight: theme.lineHeight,
    padding: condensed ? theme.tableCondensedPadding : theme.tablePadding,
    verticalAlign: 'middle',
    textAlign: align,

    // For first th/td in the first row in the first thead or tbody
    borderTopLeftRadius: tbodyIndex === 0 && rowIndex === 0 && cellIndex === 0
      ? theme.borderRadius
      : undefined,
    // The last index is passed down as its complement so you can detect whether
    // it's the last child easily while preserving its magnitude

    // For last th/td in the first row in the first thead or tbody
    borderTopRightRadius: tbodyIndex === 0 && rowIndex === 0 && cellIndex < 0
      ? theme.borderRadius
      : undefined,

    // For first th/td (can be either) in the last row in the last thead, tbody, and tfoot
    borderBottomLeftRadius: (tbodyIndex < 0 || tfootIndex < 0) &&
      rowIndex < 0 &&
      cellIndex === 0
      ? theme.borderRadius
      : undefined,

    // For last th/td (can be either) in the last row in the last thead, tbody, and tfoot
    borderBottomRightRadius: (tbodyIndex < 0 || tfootIndex < 0) &&
      rowIndex < 0 &&
      cellIndex < 0
      ? theme.borderRadius
      : undefined,

    // Clear border-radius for first and last td in the last row in the last tbody for table with tfoot
    borderRadius: tfootIndex != null &&
      tbodyIndex < 0 &&
      rowIndex < 0 &&
      (cellIndex === 0 || cellIndex < 0)
      ? 0
      : undefined,

    // Responsive styles
    tablet: {
      display: 'block',
      width: '100%',
      boxSizing: 'border-box',
      borderTopWidth: 0,
      borderTop: cellIndex === 0
        ? `1px solid ${theme.tableBordercolor}`
        : undefined
    },

    // Row Styles
    borderColor: theme[`table${capitalize(rowType)}BorderColor`],
    backgroundColor: striped && rowIndex % 2 === 1
      ? theme[`tableStriped${capitalize(tableModifier)}Color`]
      : theme[`table${capitalize(rowType)}Color`],

    ':hover': hover
      ? { backgroundColor: theme[`tableHover${capitalize(rowType)}Color`] }
      : undefined,
    borderLeft: bordered ? `1px solid ${theme.tableBorderColor}` : undefined,

    [`@media screen and (max-width: ${theme.breakpoints.tablet})`]: {
      display: 'block',
      border: 'none',
      float: 'left',
      clear: 'left',
      padding: 0,

      // Mobile stripe issue fix
      backgroundColor: striped && rowIndex % 2 === 1 ? 'transparent' : undefined
    }
  });
