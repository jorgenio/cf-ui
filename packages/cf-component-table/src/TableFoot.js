import React from 'react';
import PropTypes from 'prop-types';
import { filterStyle, mapChildren } from 'cf-style-container';

class TableFoot extends React.Component {
  render() {
    const {
      className,
      styles,
      striped,
      hover,
      bordered,
      condensed,
      bare,
      tfootIndex,
      children,
      ...props
    } = this.props;

    return (
      <tfoot
        className={className || (styles && styles.tableFoot)}
        {...filterStyle(props)}
      >
        {mapChildren(children, (child, index, children) => {
          const isLast = index === children.length - 1;
          if (React.isValidElement(child)) {
            return React.cloneElement(child, {
              condensed,
              striped,
              bordered,
              hover,
              bare,
              tfootIndex,
              rowIndex: isLast ? -index : index
            });
          }
          return child;
        })}

      </tfoot>
    );
  }
}

TableFoot.propTypes = {
  className: PropTypes.string,
  styles: PropTypes.object,
  children: PropTypes.node
};

export default TableFoot;
