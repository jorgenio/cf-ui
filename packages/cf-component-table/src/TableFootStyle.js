export default () => ({
  tablet: {
    display: 'block',
    width: '100%',
    boxSizing: 'border-box'
  }
});
