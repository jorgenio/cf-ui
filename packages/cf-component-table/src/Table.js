import React from 'react';
import PropTypes from 'prop-types';
import { filterStyle, filterNone, mapChildren } from 'cf-style-container';
import TableHead from './TableHead';
import TableBody from './TableBody';
import TableFoot from './TableFoot';

function issubclass(childType, Type) {
  return (
    childType === Type || (childType && childType.prototype instanceof Type)
  );
}

class Table extends React.Component {
  render() {
    const {
      className,
      styles,
      striped,
      hover,
      bordered,
      condensed,
      bare,
      children,
      ...props
    } = this.props;

    const sectionCount = React.Children.toArray(children).reduce((
      acc,
      child
    ) => {
      if (issubclass(child.type, TableHead)) acc.theads++;
      else if (issubclass(child.type, TableBody)) acc.tbodies++;
      else if (issubclass(child.type, TableFoot)) acc.tfoots++;
      return acc;
    }, { theads: 0, tbodies: 0, tfoots: 0 });

    const { theads, tbodies, tfoots } = sectionCount;
    let theadsSeen = 0, tbodiesSeen = 0, tfootsSeen = 0;

    return (
      <table
        className={className || (styles && styles.table)}
        {...filterStyle(props)}
      >
        {mapChildren(children, (child, index, children) => {
          if (React.isValidElement(child)) {
            if (issubclass(child.type, TableHead)) theadsSeen++;
            else if (issubclass(child.type, TableBody)) tbodiesSeen++;
            else if (issubclass(child.type, TableFoot)) tfootsSeen++;

            const isLastThead = theadsSeen == theads;
            const isLastTbody = tbodiesSeen == tbodies;
            const isLastTfoot = tfootsSeen == tfoots;

            return React.cloneElement(
              child,
              filterNone({
                condensed,
                striped,
                bordered,
                hover,
                bare,
                theadIndex: issubclass(child.type, TableHead)
                  ? (theadsSeen - 1) * (isLastThead ? -1 : 1)
                  : undefined,
                tbodyIndex: issubclass(child.type, TableBody)
                  ? (tbodiesSeen - 1) * (isLastTbody ? -1 : 1)
                  : undefined,
                tfootIndex: issubclass(child.type, TableFoot)
                  ? (tfootsSeen - 1) * (isLastTfoot ? -1 : 1)
                  : undefined
              })
            );
          }
          return child;
        })}
      </table>
    );
  }
}

Table.propTypes = {
  className: PropTypes.string,
  styles: PropTypes.object,
  striped: PropTypes.bool,
  hover: PropTypes.bool,
  bordered: PropTypes.bool,
  condensed: PropTypes.bool,
  bare: PropTypes.bool,
  children: PropTypes.node
};

Table.defaultProps = {
  striped: false,
  hover: false,
  bordered: true,
  condensed: false,
  bare: false
};

export default Table;
