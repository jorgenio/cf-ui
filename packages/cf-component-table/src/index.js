import TableUnstyled from './Table';
import TableBodyUnstyled from './TableBody';
import TableCellUnstyled from './TableCell';
import TableFootUnstyled from './TableFoot';
import TableHeadUnstyled from './TableHead';
import TableHeadCellUnstyled from './TableHeadCell';
import TableRowUnstyled from './TableRow';
import TableStyle from './TableStyle';
import TableBodyStyle from './TableBodyStyle';
import TableCellStyle from './TableCellStyle';
import TableFootStyle from './TableFootStyle';
import TableHeadStyle from './TableHeadStyle';
import TableHeadCellStyle from './TableHeadCellStyle';
import TableRowStyle from './TableRowStyle';
import TableTheme from './TableTheme';
import TablePropTypes from './TablePropTypes';

import { applyTheme, createComponentStyles } from 'cf-style-container';

const Table = applyTheme(
  createComponentStyles({ table: TableStyle }, TableUnstyled),
  TableTheme
);
const TableBody = applyTheme(
  createComponentStyles({ tableBody: TableBodyStyle }, TableBodyUnstyled),
  TableTheme
);
const TableCell = applyTheme(
  createComponentStyles({ tableCell: TableCellStyle }, TableCellUnstyled),
  TableTheme
);
const TableFoot = applyTheme(
  createComponentStyles({ tableFoot: TableFootStyle }, TableFootUnstyled),
  TableTheme
);
const TableHead = applyTheme(
  createComponentStyles({ tableHead: TableHeadStyle }, TableHeadUnstyled),
  TableTheme
);
const TableHeadCell = applyTheme(
  createComponentStyles(
    { tableHeadCell: TableHeadCellStyle },
    TableHeadCellUnstyled
  ),
  TableTheme
);
const TableRow = applyTheme(
  createComponentStyles({ tableRow: TableRowStyle }, TableRowUnstyled),
  TableTheme
);

export {
  Table,
  TableBody,
  TableCell,
  TableFoot,
  TableHead,
  TableHeadCell,
  TableRow,
  TableUnstyled,
  TableBodyUnstyled,
  TableCellUnstyled,
  TableFootUnstyled,
  TableHeadUnstyled,
  TableHeadCellUnstyled,
  TableRowUnstyled,
  TableTheme,
  TablePropTypes
};
