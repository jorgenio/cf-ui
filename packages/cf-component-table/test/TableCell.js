import React from 'react';
import renderer from 'react-test-renderer';
import { TableCell } from '../../cf-component-table/src/index';
import felaTestContext from '../../../felaTestContext';

test('should render', () => {
  const component = renderer.create(
    felaTestContext(<TableCell>TableCell</TableCell>)
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('should render extra class name', () => {
  const component = renderer.create(
    felaTestContext(<TableCell className="extra">TableCell</TableCell>)
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('should render with align', () => {
  const component = renderer.create(
    felaTestContext(<TableCell align="center">TableCell</TableCell>)
  );
  expect(component.toJSON()).toMatchSnapshot();
});
