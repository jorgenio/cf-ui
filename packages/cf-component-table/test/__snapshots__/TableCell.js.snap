// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<td
  className="cf-xm5wql"
>
  TableCell
</td>
`;

exports[`should render extra class name 1`] = `
<td
  className="extra"
>
  TableCell
</td>
`;

exports[`should render with align 1`] = `
<td
  className="cf-1dzevu8"
>
  TableCell
</td>
`;
